File: frontend/src/scripts/ingest-kb.ts
/**
 * Knowledge base ingestion script.
 * Parses src/ai/knowledge-base.md, embeds each section, and upserts into Supabase.
 *
 * Run: npm run ingest-kb
 */

import fs from "fs";
import path from "path";
import { OpenAIEmbeddings } from "@langchain/openai";
import { getAdminClient } from "@/services/supabase-admin";
import { env } from "@/config/env";

interface KBSection {
  section: string;
  content: string;
  roles: string[];
}

function parseKnowledgeBase(raw: string): KBSection[] {
  const sections: KBSection[] = [];
  // Split on lines that start with "## " (h2 headings)
  const parts = raw.split(/^(?=## )/m);

  for (const part of parts) {
    const lines = part.split("\n");
    const heading = lines[0]?.trim();
    if (!heading?.startsWith("## ")) continue;

    const section = heading.replace(/^## /, "").trim();

    // Look for <!-- roles: ... --> comment in first 3 lines after heading
    let roles: string[] = ["admin", "staff", "customer"];
    const rolesLine = lines.slice(1, 4).find((l) => l.trim().startsWith("<!-- roles:"));
    if (rolesLine) {
      const match = rolesLine.match(/<!--\s*roles:\s*([^>]+)\s*-->/);
      if (match) {
        roles = match[1]
          .split(",")
          .map((r) => r.trim())
          .filter(Boolean);
      }
    }

    // Content = everything after the heading and optional roles comment, stripped of HR separators
    const content = lines
      .slice(1)
      .filter((l) => !l.trim().startsWith("<!-- roles:") && l.trim() !== "---")
      .join("\n")
      .trim();

    if (content.length > 0) {
      sections.push({ section, content, roles });
    }
  }

  return sections;
}

async function main() {
  const kbPath = path.join(process.cwd(), "src/ai/knowledge-base.md");
  const raw = fs.readFileSync(kbPath, "utf-8");
  const sections = parseKnowledgeBase(raw);

  console.log(`Parsed ${sections.length} sections from knowledge base`);
  sections.forEach((s) => console.log(`  [${s.roles.join(",")}] ${s.section}`));

  const embeddings = new OpenAIEmbeddings({
    model: env.openai.embeddingModel,
    apiKey: env.openai.apiKey,
  });

  const supabase = getAdminClient();

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const db = supabase as any;

  // Clear existing chunks
  const { error: deleteError } = await db.from("document_chunks").delete().neq("id", 0);
  if (deleteError) {
    console.error("Failed to clear existing chunks:", deleteError.message);
    process.exit(1);
  }

  // Embed and insert each section
  for (const { section, content, roles } of sections) {
    process.stdout.write(`  Embedding: ${section} ... `);
    const vectors = await embeddings.embedDocuments([content]);
    const embedding = vectors[0];

    const { error } = await db.from("document_chunks").insert({
      section,
      content,
      embedding: JSON.stringify(embedding),
      roles,
    });

    if (error) {
      console.error(`\nFailed to insert "${section}":`, error.message);
      process.exit(1);
    }
    console.log("done");
  }

  console.log(`\nIngested ${sections.length} chunks successfully.`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: backend/scripts/ingest_kb.py
"""Knowledge base ingestion script.

Parses backend/knowledge-base.md, embeds each section, and upserts into Supabase.

Run from the backend/ directory:
    python -m scripts.ingest_kb
"""

import re
import sys
from pathlib import Path

from app.ai.models import embeddings
from app.core.supabase import get_admin_client

KB_PATH = Path(__file__).resolve().parents[1] / "knowledge-base.md"

DEFAULT_ROLES = ["admin", "staff", "customer"]


def parse_knowledge_base(raw: str) -> list[dict]:
    """Parse knowledge base markdown into sections."""
    sections: list[dict] = []
    parts = re.split(r"^(?=## )", raw, flags=re.MULTILINE)

    for part in parts:
        lines = part.split("\n")
        heading = lines[0].strip() if lines else ""
        if not heading.startswith("## "):
            continue

        section = heading[len("## ") :].strip()

        roles = list(DEFAULT_ROLES)
        roles_line = next(
            (line for line in lines[1:4] if line.strip().startswith("<!-- roles:")),
            None,
        )
        if roles_line:
            match = re.search(r"<!--\s*roles:\s*([^>]+)\s*-->", roles_line)
            if match:
                roles = [r.strip() for r in match.group(1).split(",") if r.strip()]

        content = "\n".join(
            line
            for line in lines[1:]
            if not line.strip().startswith("<!-- roles:") and line.strip() != "---"
        ).strip()

        if content:
            sections.append({"section": section, "content": content, "roles": roles})

    return sections


def main() -> None:
    """Main ingestion function."""
    raw = KB_PATH.read_text(encoding="utf-8")
    sections = parse_knowledge_base(raw)

    print(f"Parsed {len(sections)} sections from knowledge base")
    for s in sections:
        print(f"  [{','.join(s['roles'])}] {s['section']}")

    supabase = get_admin_client()

    delete_res = supabase.table("document_chunks").delete().neq("id", 0).execute()
    if getattr(delete_res, "error", None):
        print(f"Failed to clear existing chunks: {delete_res.error}")
        sys.exit(1)

    for s in sections:
        section, content, roles = s["section"], s["content"], s["roles"]
        print(f"  Embedding: {section} ... ", end="", flush=True)
        vectors = embeddings.embed_documents([content])
        embedding = vectors[0]

        res = (
            supabase.table("document_chunks")
            .insert(
                {
                    "section": section,
                    "content": content,
                    "embedding": embedding,
                    "roles": roles,
                }
            )
            .execute()
        )
        if getattr(res, "error", None):
            print(f"\nFailed to insert \"{section}\": {res.error}")
            sys.exit(1)
        print("done")

    print(f"\nIngested {len(sections)} chunks successfully.")


if __name__ == "__main__":
    try:
        main()
    except Exception as err:
        print(err)
        sys.exit(1)
